### PR TITLE
Added stub for sre_compile(2and3)

### PR DESCRIPTION
**Pull request by ashwch**

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
Can you fix the Travis failure? Looks like you may have to write stubs for `sre_parse` first.

**Comment by ashwch** [comment]
Will do it.

**Comment by ashwch** [comment]
@JelleZijlstra Added stubs for `sre_parse` and `sre_constants` for Python 2 and 3.

- https://github.com/python/typeshed/pull/1500
- https://github.com/python/typeshed/pull/1502

File: stdlib/2and3/sre_compile.pyi [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+import sys
+from sre_parse import SubPattern
+from typing import Any, Pattern, Tuple, Type, Union
+
+MAXCODE = ...  # type: int
+if sys.version_info < (3, 0):
+    STRING_TYPES = ...  # type: Union[Tuple[Type[str]], Tuple[Type[str], Type[unicode]]]
+
+def isstring(obj: Any) -> int: ...
```
Comment (JelleZijlstra) [inline stdlib/2and3/sre_compile.pyi]: The argument type should be `object` and the return type should be `bool` (at least in Python 3).

File: stdlib/2and3/sre_compile.pyi [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+def compile(p: Union[str, SubPattern], flags: int = ...) -> Pattern: ...
```
Comment (JelleZijlstra) [inline stdlib/2and3/sre_compile.pyi]: I think it also accepts `bytes` in Python 3, and will return either a `Pattern[str]` or a `Pattern[bytes]` depending on its argument type.

File: stdlib/2and3/sre_compile.pyi [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+# and https://github.com/python/cpython/blob/master/Lib/sre_compile.py
+
+import sys
+from sre_parse import SubPattern
+from typing import Any, Pattern, Tuple, Type, Union
+
+MAXCODE = ...  # type: int
+if sys.version_info < (3, 0):
+    STRING_TYPES = ...  # type: Union[Tuple[Type[str]], Tuple[Type[str], Type[unicode]]]
+
```
Comment (JelleZijlstra) [inline stdlib/2and3/sre_compile.pyi]: Missing `sre_compile.dis` in Python 3.

File: stdlib/2and3/sre_compile.pyi [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# Source: https://hg.python.org/cpython/file/2.7/Lib/sre_compile.py
+# and https://github.com/python/cpython/blob/master/Lib/sre_compile.py
+
+import sys
+from sre_parse import SubPattern
+from typing import Any, Pattern, Tuple, Type, Union
+
+MAXCODE = ...  # type: int
+if sys.version_info < (3, 0):
+    STRING_TYPES = ...  # type: Union[Tuple[Type[str]], Tuple[Type[str], Type[unicode]]]
```
Comment (JelleZijlstra) [inline stdlib/2and3/sre_compile.pyi]: I think we don't need to worry about the case where `unicode` does not exist. You can just say `STRING_TYPES = (str, unicode)`, similar to what the stubs for `six` do.

File: stdlib/2and3/sre_compile.pyi (hunk 1 above)
Comment (ashwch) [inline stdlib/2and3/sre_compile.pyi]: Umm... why not `Any`?

Comment (JelleZijlstra) [inline stdlib/2and3/sre_compile.pyi]: `Any` is intended to mean that the type system isn't powerful enough to encode the actual type, and `object` means that the function genuinely accepts every kind of Python object.

File: stdlib/2and3/sre_compile.pyi (hunk 2 above)
Comment (ashwch) [inline stdlib/2and3/sre_compile.pyi]: Considering it also accepts `SubPattern`, I am not sure how we can specify a specific return type in that case.

**Comment by ashwch** [comment]
@JelleZijlstra Any idea about the error: https://travis-ci.org/python/typeshed/jobs/262318784 ?

**Pull request merged**